src/EditorFeatures/CSharpTest/UsePatternCombinators/CSharpUsePatternCombinatorsDiagnosticAnalyzerTests.cs

```diff
@@ -0,0 +1,111 @@
+    bool Method() => EXPRESSION;
+    bool Prop1 => EXPRESSION;
+    bool Prop2 { get } = EXPRESSION;
+    void If() { if (EXPRESSION) ; }
+    void Argument1() => Test(EXPRESSION);
+    void Argument2() => Test(() => EXPRESSION);
+    void Argument3() => Test(_ => EXPRESSION);
+    void For() { for (; EXPRESSION; ); }
+    void Local() { var local = EXPRESSION; }
+    void Assignment() { _ = EXPRESSION; }
+    void Do() { do ; while (EXPRESSION); }
+    void While() { while (EXPRESSION) ; }
+    void When() { _ = o switch { _ when EXPRESSION => 0 }; }
+    bool Return() { return EXPRESSION; }
+    IEnumerable<bool> YieldReturn() { yield return EXPRESSION; }
+    Func<object, bool> SimpleLambda() => o => EXPRESSION;
+    Func<bool> ParenthesizedLambda() => () => EXPRESSION;
+}
+";
+
+        [InlineData("i == 0")]
+        [InlineData("i > 0")]
+        [InlineData("i is C")]
+        [InlineData("i is C c")]
+        [InlineData("!(i > 0)")]
+        [InlineData("o != null")]
+        [Theory, Trait(Traits.Feature, Traits.Features.CodeActionsUsePatternCombinators)]
+        public async Task TestMissingOnExpression(string expression)
+        {
+            await TestMissingAsync(s_initialMarkup.Replace("EXPRESSION", expression));
+        }
+
+        [InlineData("!(o is C c)", "o is not C c")]
+        [InlineData("!(o is C)", "o is not C")]
+        [InlineData("i == 1 || 2 == i", "i is 1 or 2")]
+        [InlineData("i != 1 || 2 != i", "i is not 1 or not 2")]
+        [InlineData("i != 1 && 2 != i", "i is not (1 or 2)")]
+        [InlineData("!(i != 1 && 2 != i)", "i is 1 or 2")]
+        [InlineData("i < 1 && 2 <= i", "i is < 1 and >= 2")]
+        [InlineData("i < 1 && 2 <= i && i is not 0", "i is < 1 and >= 2 and not 0")]
+        [InlineData("(int.MaxValue - 1D) < i && i > 0", "i is > (int.MaxValue - 1D) and > 0")]
+        [Theory, Trait(Traits.Feature, Traits.Features.CodeActionsUsePatternCombinators)]
+        public async Task TestOnExpression(string expression, string expected)
+        {
+            await TestInRegularAndScriptAsync(
+                s_initialMarkup.Replace("EXPRESSION", expression),
+                s_initialMarkup.Replace("EXPRESSION", expected));
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUsePatternCombinators)]
+        public async Task TestMultiline()
+        {
+            await TestInRegularAndScript1Async(
+@"class C
+{
+    bool M0(int variable)
+    {
+        return {|FixAllInDocument:variable == 0 ||
+               variable == 1 ||
+               variable == 2|};
+    }
+    bool M1(int variable)
+    {
+        return variable != 0 &&
+               variable != 1 &&
+               variable != 2;
+    }
+}",
+@"class C
+{
+    bool M0(int variable)
+    {
+        return variable is 0 or
+               1 or
+               2;
+    }
+    bool M1(int variable)
+    {
+        return variable is not (0 or
+               1 or
+               2);
```

Comment:
same earlier comment that we should think about reducing extremely simple exprs left over.  i.e. if we're down to basiclly a primary expression on a single line, we might want to collapse that.

Reply:
>if we're down to basiclly a primary expression on a single line, we might want to collapse that.

I don't undrestand what this means.

My preference is to not do any formatting stuff in the feature (except for keeping trivia around). I believe those belong to the formatter which runs in an independant pass.

Reply:
>if we're down to basiclly a primary expression on a single line,

Ok I think now I read it correctly. Shouldn't we use that heuristic to format multiline expressions everywhere?

Reply:
i would be interested to see what that looks liek.  note: i think it's not as simple as "a primary expression".  For example, `new Honking<Big, Type>(Honking, Big, Expression)` is technically a primary expression.

I think i'm looking for some notion of 'simple'.  Which, to me, would likely start by being limited to expressions that are:
1. definitely only on a single line. 
2. probably only built out of literals, or types.  i.e.  `0`, or `A.B.C`
3. likely without non-whitespace trivia (which slightly falls out of '1')

Basically, something htat feels practically 'atomic'.  Keeping these atoms on a single line feels weird to me, and i think in those cases it would possible be appropriate to munge with trivia.  

It would be good to test out and see the results on roslyn to see how it feels. Thanks!